### PR TITLE
feat: `pragma-utils` typescript + Deployment refactor + Every scripts

### LINKED ISSUE #59: dev: Oracle scripts
# Current Behavior

We are able to deploy our contracts on any starknet & evm chains.
However, there is no simple way to add new currencies, pair or publishers.

# Expected Behavior

We should have some scripts in the `pragma-scripts`  package allowing us to:
#### pragma-oracle
* uprade-pragma-oracle
* add-pair
* add-currency
* register-publisher
* add-source
* remove-source
* remove-publisher

#### pragma.sol
* upgrade-pragma

For reference, see our pragma-oracle v1 repository:
https://github.com/astraly-labs/pragma-oracle/tree/main/pragma-deployer

### PR DESCRIPTION
**Pull request by akhercha**
Resolves: #59

## Work done
* New pragma-utils package with our utilities
* All scripts!
* Refactored the deployment scripts

### CODE REVIEW

File: typescript/pragma-scripts/README.md [hunk 1]
```diff
@@ -18,14 +18,120 @@ bun install
...
 
-### Usage
+Located in the `dispatcher/` directory.
+
+#### add_feeds.ts
+
+Adds multiple data feeds to the Pragma Dispatcher contract.
+
+```bash
+bun run dispatcher/add_feeds.ts --chain <chain_name> --feed-ids <feed_id1> <feed_id2> ...
```
Comment (EvolveArt) [inline typescript/pragma-scripts/README.md]: update that to dispatcher:add_feeds as done in package.json

File: typescript/pragma-scripts/package-lock.json [hunk 2]
```diff
@@ -10,7 +10,9 @@
         "commander": "^12.1.0",
         "dotenv": "^16.4.5",
         "ethers": "^6.13.3",
-        "js-yaml": "^4.1.0"
+        "js-yaml": "^4.1.0",
```
Comment (EvolveArt) [inline typescript/pragma-scripts/package-lock.json:13]: remove file

File: typescript/pragma-scripts/src/pragma/update_feed.ts [hunk 3]
```diff
@@ -122,13 +114,19 @@ async function getCalldataFromTheoros(
 async function main() {
   const options = parseCommandLineArguments();
 
-  const chainConfig = getChainConfig(options.targetChain);
+  const chain = options.chain;
```
Comment (EvolveArt) [inline typescript/pragma-scripts/src/pragma/update_feed.ts:91]: remove unusued getChainConfig function

File: typescript/pragma-scripts/src/pragma/update_feed.ts [hunk 4]
```diff
@@ -41,17 +42,13 @@ function parseCommandLineArguments(): OptionValues {
...
 
   program
-    .name("update-data-feed")
-    .description("CLI to update a Pragma data feed")
+    .name("update-feed")
+    .description("CLI to update a Pragma feed")
     .requiredOption(
-      "--target-chain <chain_name>",
+      "--chain <chain_name>",
       "Name of the target chain (e.g., ethereum_mainnet)",
```
Comment (EvolveArt) [inline typescript/pragma-scripts/src/pragma/update_feed.ts:23]: example invalid

File: typescript/pragma-scripts/src/oracle/get_all_publishers.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+      `Error: Invalid starknet chain, must be in ${STARKNET_CHAINS.toString()}.`,
+    );
+    process.exit(1);
+  }
+  return options;
+}
+
+async function getAllPublishersAndSources(
+  publisherRegistry: Contract,
+  account: Deployer,
```
Comment (EvolveArt) [inline typescript/pragma-scripts/src/oracle/get_all_publishers.ts]: unused arg

File: typescript/pragma-scripts/src/dispatcher/add_feeds.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+) {
+  try {
+    console.log(`\t ⏳ Adding feed ${feedId}...`);
+    const invoke = await pragmaDispatcher.invoke("add_feed", [feedId]);
+    await account.waitForTransaction(invoke.transaction_hash);
+
+    const receipt = await account.getTransactionReceipt(
+      invoke.transaction_hash,
+    );
+    await ensureSuccess(receipt, account.provider);
```
Comment (EvolveArt) [inline typescript/pragma-scripts/src/dispatcher/add_feeds.ts:49]: this code block could be extracted and replaced everywhere its used, also make sure to repeatly call getTransactionReceipt as receipt might not directly be available

File: typescript/pragma-utils/src/starknet/gas.ts [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+          console.log(
+            `Usage: append either --write or --check to the CLI command.`,
+          );
+        }
+      });
+    },
+  };
+}
+
+function shortenHex(hex: string) {
```
Comment (EvolveArt) [inline typescript/pragma-utils/src/starknet/gas.ts]: should not be here + add jsdoc

File: typescript/pragma-utils/src/starknet/gas.ts [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+    storageDiffs,
+    daMode: blockInfo.l1_da_mode,
+    calldataGas,
+    eventGas,
+  };
+}
+
+type Profile = Awaited<ReturnType<typeof profileGasUsage>>;
+
+export function newProfiler(provider: RpcProvider) {
```
Comment (EvolveArt) [inline typescript/pragma-utils/src/starknet/gas.ts:176]: where is that used ?

File: typescript/pragma-scripts/package-lock.json (hunk 2 above)
Comment (akhercha) [reply typescript/pragma-scripts/package-lock.json:13]: the lockfile? why?

File: typescript/pragma-scripts/README.md (hunk 1 above)
Comment (akhercha) [reply typescript/pragma-scripts/README.md]: done

**Review by EvolveArt — COMMENTED** [review]
> the lockfile? why? 

we already have bun lock

File: typescript/pragma-utils/src/starknet/gas.ts (hunk 8 above)
Comment (akhercha) [reply typescript/pragma-utils/src/starknet/gas.ts:176]: Nowhere yet - just grabbed it from github, might be useful *one day*

**Comment by akhercha** [comment]
> we already have bun lock

@EvolveArt  Ah, we need both in order to be compatible with `npm` - ppl should be able to run the package manager of their choices imo

File: typescript/pragma-scripts/src/dispatcher/add_feeds.ts (hunk 6 above)
Comment (akhercha) [reply typescript/pragma-scripts/src/dispatcher/add_feeds.ts:49]: it is called after `waitForTransaction` so should be gud?

File: typescript/pragma-scripts/src/pragma/update_feed.ts (hunk 4 above)
Comment (akhercha) [reply typescript/pragma-scripts/src/pragma/update_feed.ts:23]: done

File: typescript/pragma-scripts/src/oracle/get_all_publishers.ts (hunk 5 above)
Comment (akhercha) [reply typescript/pragma-scripts/src/oracle/get_all_publishers.ts]: oops, removed

File: typescript/pragma-scripts/src/pragma/update_feed.ts (hunk 3 above)
Comment (akhercha) [reply typescript/pragma-scripts/src/pragma/update_feed.ts:91]: done

File: typescript/pragma-utils/src/starknet/gas.ts (hunk 7 above)
Comment (akhercha) [reply typescript/pragma-utils/src/starknet/gas.ts]: done

File: typescript/pragma-scripts/src/dispatcher/add_feeds.ts (hunk 6 above)
Comment (EvolveArt) [reply typescript/pragma-scripts/src/dispatcher/add_feeds.ts:49]: yea should be indeed

File: typescript/pragma-utils/src/starknet/gas.ts (hunk 8 above)
Comment (EvolveArt) [reply typescript/pragma-utils/src/starknet/gas.ts:176]: not a huge fan of unused code but ok

**Review by EvolveArt — APPROVED** [review]

**Pull request merged**